### PR TITLE
Dashboard - report contract data

### PR DESCRIPTION
**Pull request by ping-ke**
related issue: https://github.com/ethstorage/es-node/issues/43
Add a simple server to report contract metrics data to DB and show it in Grafana.

How to start:
`./dashboard --address <your address> --l1contract 0x703848F4c85f18e3acd8196c8eC91eb0b7Bd0797`

How to check: 
Check the log from the address `http://<your address>:8300/metrics`

Report:
http://grafana.quarkchain.io/d/fdfcc57c-30fb-424e-8dc5-b0cf7bcce84b/ethstorage-monitoring?orgId=2

### CODE REVIEW

File: cmd/dashboard/main.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
+package main
```
Comment (qzhodl) [inline cmd/dashboard/main.go:1]: Please keep in mind that we need to add license header for every new file.

**Comment by qzhodl** [comment]
It seems that the status shown in the dashboard [here](http://65.109.63.154/d/fdfcc57c-30fb-424e-8dc5-b0cf7bcce84b/ethstorage-monitoring?orgId=2&from=1699434636210&to=1699438236210) is different with what we expected, we may need to have some discussion on it.

File: cmd/dashboard/main.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+	log.Root().SetHandler(log.LvlFilterHandler(log.Lvl(*logFlag), log.StreamHandler(os.Stderr, log.TerminalFormat(true))))
+
+	if *portFlag < 0 || *portFlag > math.MaxUint16 {
+		log.Crit("Invalid port")
+	}
+	d, err := newDashboard(*rpcURLFlag, common.HexToAddress(*l1ContractFlag))
+	if err != nil {
+		log.Crit("New dashboard fail", "err", err)
+	}
+	l1LatestBlockSub := eth.PollBlockChanges(d.ctx, d.logger, d.l1Source, d.RefreshContractMetrics, ethRPC.LatestBlockNumber,
```
Comment (qzhodl) [inline cmd/dashboard/main.go:127]: If the dashboard crashes or stops for some time, we may need to start monitoring the data from where it last stopped

File: cmd/dashboard/main.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+		log.Warn("Refresh contract metrics (last kv index) failed", "err", err.Error())
+		return
+	}
+	maxShardIdx := lastKVIndex / d.kvEntries
+	d.m.SetLastKVIndexAndMaxShardId(sig.Number, lastKVIndex, maxShardIdx)
+	d.logger.Info("RefreshContractMetrics", "lastKvIndex", lastKVIndex, "maxShardIdx", maxShardIdx)
+
+	l1api := miner.NewL1MiningAPI(d.l1Source, d.logger)
+	for sid := uint64(0); sid <= maxShardIdx; sid++ {
+		info, err := l1api.GetMiningInfo(ctx, d.l1Contract, sid)
```
Comment (qzhodl) [inline cmd/dashboard/main.go:82]: We may need to filter the mining event to report who is the miner and how much is the reward

File: cmd/dashboard/main.go [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+func (d *dashboard) RefreshContractMetrics(ctx context.Context, sig eth.L1BlockRef) {
+	lastKVIndex, err := d.l1Source.GetStorageLastBlobIdx(int64(sig.Number))
+	if err != nil {
+		log.Warn("Refresh contract metrics (last kv index) failed", "err", err.Error())
+		return
+	}
+	maxShardIdx := lastKVIndex / d.kvEntries
+	d.m.SetLastKVIndexAndMaxShardId(sig.Number, lastKVIndex, maxShardIdx)
```
[the remark this answers is not in the document]
Comment (ping-ke) [reply cmd/dashboard/main.go:77]: yes

File: cmd/dashboard/main.go [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+	if *portFlag < 0 || *portFlag > math.MaxUint16 {
+		log.Crit("Invalid port")
+	}
+	d, err := newDashboard(*rpcURLFlag, common.HexToAddress(*l1ContractFlag))
+	if err != nil {
+		log.Crit("New dashboard fail", "err", err)
+	}
+
+	l1LatestBlockSub := eth.PollBlockChanges(d.ctx, d.logger, d.l1Source, d.RefreshBlobsMetrics, ethRPC.LatestBlockNumber,
+		10*time.Second, time.Second*10)
```
Comment (qzhodl) [inline cmd/dashboard/main.go:267]: For the internal, maybe 12s is more reasonable since 1 epoch == 12s?

File: cmd/dashboard/main.go [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+		// TODO: update when new version contract deployed, use the reward in the new MinedBlock event
+		balance, nErr := d.l1Source.BalanceAt(d.ctx, d.l1Contract, new(big.Int).SetUint64(l.BlockNumber))
+		if nErr != nil {
+			return nil, start, fmt.Errorf("BalanceAt fail, block: %d, error: %s", l.BlockNumber, err.Error())
+		}
+		balanceBefore, oErr := d.l1Source.BalanceAt(d.ctx, d.l1Contract, new(big.Int).SetUint64(l.BlockNumber-1))
+		if oErr != nil {
+			return nil, start, fmt.Errorf("BalanceAt fail, block: %d, error: %s", l.BlockNumber-1, err.Error())
+		}
+		reward := new(big.Int).Sub(balanceBefore, balance)
```
Comment (qzhodl) [inline cmd/dashboard/main.go:203]: should be 99% * reward because 1% goes to the treasury

File: cmd/dashboard/main.go [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+	d, err := newDashboard(*rpcURLFlag, common.HexToAddress(*l1ContractFlag))
+	if err != nil {
+		log.Crit("New dashboard fail", "err", err)
+	}
+
+	l1LatestBlockSub := eth.PollBlockChanges(d.ctx, d.logger, d.l1Source, d.RefreshBlobsMetrics, ethRPC.LatestBlockNumber,
+		10*time.Second, time.Second*10)
+	defer l1LatestBlockSub.Unsubscribe()
+
+	go d.FilterEventAndReport()
```
Comment (qzhodl) [inline cmd/dashboard/main.go:270]: I am not sure why we have two same interval timer here (one for RefreshBlobsMetrics, one for FilterEventAndReport ), can we combine into one?

**Review by qzhodl — APPROVED** [review]

**Pull request merged**